relay-general/src/pii/processor.rs

```diff
@@ -139,6 +140,41 @@ impl<'a> Processor for PiiProcessor<'a> {
     ) -> ProcessingResult {
         process_pairlist(self, value, state)
     }
+
+    fn process_user(
+        &mut self,
+        user: &mut User,
+        _meta: &mut Meta,
+        state: &ProcessingState<'_>,
+    ) -> ProcessingResult {
+        let ip_was_valid = user.ip_address.value().map_or(true, IpAddr::is_valid);
+
+        // Recurse into the user and does PII processing on fields.
+        user.process_child_values(self, state)?;
+
+        let has_other_fields = user.id.value().is_some()
+            || user.username.value().is_some()
+            || user.email.value().is_some();
+
+        let ip_is_still_valid = user.ip_address.value().map_or(true, IpAddr::is_valid);
+
+        // If the IP address has become invalid as part of PII processing, we move it into the user
+        // ID. That ensures people can do IP hashing and still have a correct users-affected count.
+        //
+        // Right now both Snuba and EventUser discard unparseable IPs for indexing, and we assume
+        // we want to keep it that way.
+        //
+        // If there are any other fields set that take priority over the IP for uniquely
+        // identifying a user (has_other_fields), we do not want to do anything. The value will be
+        // wiped out in renormalization anyway.
+        if ip_was_valid && !has_other_fields && !ip_is_still_valid {
+            let mut user_ip = Annotated::empty();
+            mem::swap(&mut user.ip_address, &mut user_ip);
+            user.id = user_ip.map_value(IpAddr::into_inner).map_value(From::from);
```

Comment:
You can simplify this with:

```suggestion
            user.id = mem::take(&mut user.ip_address)
                .map_value(|ip| ip.into_inner().into());
```